pkg/apis/apps/kfdef/v1beta1/application_types.go

```diff
@@ -0,0 +1,170 @@
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   KfDefSpec   `json:"spec,omitempty"`
+	Status KfDefStatus `json:"status,omitempty"`
+}
+
+type KfDefSpec struct {
+	Applications []Application `json:"applications,omitempty"`
+	Plugins      []Plugin      `json:"plugins,omitempty"`
+	Secrets      []Secret      `json:"secrets,omitempty"`
+	Repos        []Repo        `json:"repos,omitempty"`
+}
+
+// Application defines an application to install
+type Application struct {
+	Name            string           `json:"name,omitempty"`
+	KustomizeConfig *KustomizeConfig `json:"kustomizeConfig,omitempty"`
+}
+
+type KustomizeConfig struct {
+	RepoRef    *RepoRef    `json:"repoRef,omitempty"`
+	Overlays   []string    `json:"overlays,omitempty"`
+	Parameters []NameValue `json:"parameters,omitempty"`
+}
+
+type RepoRef struct {
+	Name string `json:"name,omitempty"`
+	Path string `json:"path,omitempty"`
+}
+
+type NameValue struct {
+	Name         string `json:"name,omitempty"`
+	Value        string `json:"value,omitempty"`
+	InitRequired bool   `json:"initRequired,omitempty"`
+}
+
+// Plugin can be used to customize the generation and deployment of Kubeflow
+type Plugin struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec *runtime.Object `json:"spec,omitempty"`
+}
+
+// Secret provides information about secrets needed to configure Kubeflow.
+// Secrets can be provided via references e.g. a URI so that they won't
+// be serialized as part of the KfDefSpec which is intended to be written into source control.
+type Secret struct {
+	Name         string        `json:"name,omitempty"`
+	SecretSource *SecretSource `json:"secretSource,omitempty"`
+}
+
+type SecretSource struct {
+	LiteralSource *LiteralSource `json:"literalSource,omitempty"`
+	HashedSource  *HashedSource  `json:"hashedSource,omitempty"`
+	EnvSource     *EnvSource     `json:"envSource,omitempty"`
+}
+
+type LiteralSource struct {
+	Value string `json:"value,omitempty"`
+}
+
+type HashedSource struct {
+	HashedValue string `json:"value,omitempty"`
+}
+
+type EnvSource struct {
+	Name string `json:"Name,omitempty"`
+}
+
+// Repo provides information about a repository providing config (e.g. kustomize packages,
+// Deployment manager configs, etc...)
+type Repo struct {
+	// Name is a name to identify the repository.
+	Name string `json:"name,omitempty"`
+	// URI where repository can be obtained.
+	// Can use any URI understood by go-getter:
+	// https://github.com/hashicorp/go-getter/blob/master/README.md#installation-and-usage
+	Uri string `json:"uri,omitempty"`
+}
+
+// KfDefStatus defines the observed state of KfDef
+type KfDefStatus struct {
+	Conditions []KfDefCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,6,rep,name=conditions"`
+	// ReposCache is used to cache information about local caching of the URIs.
+	ReposCache map[string]RepoCache `json:"reposCache,omitempty"`
```

Comment:
@gabrielwen @yanniszark Should we make ReposCache a list instead of a Map and move name into RepoCache?

This has two advantages

1. RepoCache would be more consistent with Repo
1. RepoCache would contain all the relevant information about a cached repo. 
    * Right now the name is stored as the key in the Map. So you can't just pass around a RepoCache struct; you'd have to pass around the tuple (name, RepoCache).

Reply:
I'm fine with using a list.  updated the definition.

Reply:
@jlewi this is more in line with the Kubernetes API Conventions.
I'm also in favor of this.